Add custom datalist component

LMS007 opened the pull request:
- Rework tag-editor to use new custom datalist
- Remove tag-editor’s Chrome hack fix to limit datalist to 20
- useElementShouldClose now works on Preact component as well as native components
![datalist](https://user-images.githubusercontent.com/3939074/72045436-ed74d300-326a-11ea-8b88-3707a7b3febb.gif)


Also fixes https://github.com/hypothesis/client/issues/1690

LMS007 commented on src/sidebar/components/hooks/use-element-should-close.js:
```diff
@@ -42,7 +42,11 @@ function useElementShouldClose(closeableEl, isOpen, handleClose) {
       event => {
-        if (!closeableEl.current.contains(event.target)) {
+        // preact element or just a dom element?
+        const node = closeableEl.current.base
```
This was not working out of the box on a top level preact component. From my research the "base" object is a solid test to check

robertknight replied:
I'm not sure if the `base` property of Preact components is considered part of its public API. I've asked on their Slack channel for confirmation.

In any case, as long as you're passing a ref to a DOM element in, I don't think this should be necessary. If you do think it is, can you clarify why?

LMS007 commented on src/sidebar/components/tag-editor.js:
```diff
@@ -1,30 +1,28 @@
-const SvgIcon = require('./svg-icon');
 
-// Global counter used to create a unique id for each instance of a TagEditor
-let datalistIdCounter = 0;
```
The chrome fix is all removed now

LMS007 commented on src/sidebar/components/tag-editor.js:
```diff
@@ -87,96 +95,90 @@ function TagEditor({ onEditTags, tags: tagsService, tagList }) {
-    }
-  };
-
   const handleOnInput = e => {
```
handleOnInput is now very simple and is only checking for changes to the text value in the input field, the other stuff is now handled all in keyDown because we have to have keyDown handler to check for up/down key navigation anyway and it just made more sense to move the "Enter" AND "," checking there to add a tag.

LMS007 replied:
handleOnInput is now very simple and is only checking for changes to the text value in the input field, the other stuff is now handled all in handleKeyDown because we have to have a keyDown handler to check for up/down key selection navigation anyway, and it just made more sense to move the "Enter" AND "," delimiter checking there to add a new tag.

LMS007 commented on src/sidebar/components/tag-editor.js:
```diff
@@ -87,96 +95,90 @@ function TagEditor({ onEditTags, tags: tagsService, tagList }) {
+   *  This will add a new tag.
+   */
+  const handleSelect = item => {
+    if (item) {
```
we could get away w/o this condition, but addTag does not check for empty objects, I could go either way on this

LMS007 commented on src/sidebar/components/tag-editor.js:
```diff
@@ -206,17 +208,25 @@ function TagEditor({ onEditTags, tags: tagsService, tagList }) {
+          onKeyDown={handleKeyDown}
+          onFocus={handleFocus}
+          ref={inputEl}
+          placeholder="Add new tags"
```
I changed this value because when testing a11y it read this out with the word "ellipsis" and that sucks! I could not figure out how to get it to not read this so I figured I would just make it work for screen readers in the first place.

robertknight replied:
I guess if you wanted to have the visual and verbal text be different you'd need to set an `aria-label` property. The solution you've gone for here seems fine to me though.

jaredpdesigns replied:
I agree with @robertknight in terms of replacing the `placeholder` with something that doesn’t conflict with a11y issues.

LMS007 commented on src/sidebar/components/tag-editor.js:
```diff
@@ -206,17 +208,25 @@ function TagEditor({ onEditTags, tags: tagsService, tagList }) {
-        type="text"
-      />
-      {suggestionsList()}
+      <span ref={closeWrapperRef}>
```
I needed an extra node here for the close ref so it would work outside both the datalist and the input

robertknight commented on src/styles/sidebar/components/datalist.scss:
```diff
@@ -0,0 +1,41 @@
+  position: absolute;
+  width: 100%;
+  padding: 0;
+  background-color: var.$grey-mid;
```
I guess you are using a dark theme locally and have gone with colors to match that? So this fits in with whatever the user's configuration is locally, I was going to suggest that the default colors assume a light theme and then we use `prefers-color-scheme` media queries in CSS [to adapt to Dark Mode](https://css-tricks.com/dark-modes-with-css/). Does that sound reasonable?

We can do this adaptation on a per-component basis and potentially in future we could also use CSS variables for application-wide color changes.

jaredpdesigns replied:
@robertknight @LMS007 I could certainly help provide appropriate CSS for dark-theming—it’s already built into the Pattern Library stuff however

LMS007 replied:
Yes, I originally had all that in the Datalist, but moved it out as I needed the keyboard events to work off the input, but I think I can change the props around and move the logic for the carousel navigation into the DL component

LMS007 replied:
@jaredpdesigns , yes I think two styles here would be good, a light and dark theme. Pretty much the only css is a bit for the little arrow when the list is open and then the highlight colors for keyboard navigation and a slightly lighter color for mouse hover. But I'm open to anything here.

I also don't have any transitions. IF we added a fade transition, it gets . a bit more complicated with state and re-flow, and I could do that on a second PR.

robertknight commented on src/sidebar/components/datalist.js:
```diff
@@ -0,0 +1,73 @@
+          })}
+          onClick={() => {
+            onSelectItem(item);
+          }}
```
We'll need appropriate ARIA properties to associate the datalist with the input and to indicate the currently selected item in the datalist. I don't know the exact details of the preferred pattern for that. Can you look into it?

LMS007 replied:
Sorry, didn't add too many details to this PR last night. I did want to mention I struggled a bit with this as it would end up reading everything in the list and got crazy when I had 50 items. I also observed that the original `<datalist>` was not being read by at least Chrome Vox so I would like to know what the aural  behavior should be here. I'll dig into this more!

LMS007 commented:
- Added all the necessary tests (so far)
- Noticed one issue with hiding the list when the input is empty and fixed that
- I did not touch a11y _yet_, I added some suggestions in slack here https://hypothes-is.slack.com/archives/C1M8NH76X/p1578680573071300 to be further discussed

I did take a look at moving the up/down navigation into Datalist, but struggled because the events are triggered from the input and passing those events seems less clean than just passing which index is active. Furthermore, the parent component needs to know which index is active because it also listens for `Enter` key and needs to pick that out of the original list. It just seemed more correct to keep everything it the parent. Does that does make tag-editor heavy weight and datalist light, but I think that implementation proved to be easier at least the way I was thinking about it.


TODO: css theming.

LMS007 commented:
The new css is designed to look more like the Menu component, but with some small changes. I decided not to use an actual icon but keep the css triangle as that was a bit more flexible and simple. I also just kept the triangle a solid color as that was easier to deal with.

Hover color in dark mode lightens the background, rather than darkest it. I think that is inverse of how our primary buttons work, but I find that more intuitive and it is how the original  datalist works too.

There is still not changes to a11y. I think about the only useful thing we could do here is my suggestion in slack #frontend to make the input value bind to the datalist selected value when navigating down down the list. In that case, a reader would inform the user what the option is, but I have not implemented that here.

![datalist2-light](https://user-images.githubusercontent.com/3939074/72389310-ec192f80-36dc-11ea-8401-ba4ad92bf2b0.gif)

![datalist2](https://user-images.githubusercontent.com/3939074/72389311-ec192f80-36dc-11ea-87e9-303ba48d3583.gif)

jaredpdesigns commented:
@LMS007 I played around with styling a pseudo-element for the triangle as a CSS-only solution and like the results: https://codepen.io/jaredpdesigns/pen/WNbKERb?editors=1100#0

Pertinent CSS below:
``` scss
&:before {
  background-color: inherit;
  border: inherit;
  clip-path: polygon(0 0, 100% 0, 0% 100%, 0% 100%);
  content: "";
  display: inline-block;
  height: 0.5rem;
  left: 0.5rem;
  position: absolute;
  top: -0.3125rem;
  transform: rotate(45deg);
  width: 0.5rem;
}
```

LMS007 commented:
Great suggestion, I like this css a lot more

![Uploading Screen Shot 2020-01-14 at 5.05.03 PM.png…]()
![Screen Shot 2020-01-14 at 5 05 10 PM](https://user-images.githubusercontent.com/3939074/72396361-47edb380-36f1-11ea-9a96-2c25472b1ff5.png)

LMS007 commented:
Added accessibility based on w3 model. "Combobox with Listbox Popup Examples"
https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html

This adds a few links that require a unique id both on the ul and then again on the li tags so there was a number of additional props needed to be passed down to the Datalist.

I tested this with ChromeVox and Narrator and it does readout the results on keyboard navigation which is what we wanted to achieve. I think this implementation is still a bit unique in that we clear out the suggestion on enter and I'm not sure how obvious that will be with a screen reader.  Other then that, I think we are a lot closer to a final product here.

LMS007 commented on src/sidebar/components/datalist.js:
```diff
@@ -0,0 +1,100 @@
+  /**
+   * Optional unique HTML attribute id.
+   */
+  id: propTypes.string,
```
In an effort to make Datalist somewhat decoupled from TagEditor, I decided to make id and itemPrefixId optional

robertknight commented on src/styles/sidebar/components/datalist.scss:
```diff
@@ -0,0 +1,75 @@
+  padding: 0;
+  > li {
+    padding: 0.25em 1.5em 0.25em 1em;
+    border-left: 4px transparent solid;
```
On mobile the list items are quite short vertically if you are trying to tap them. Can you add a `@media (pointer: coarse) { ... }` query to make sure that the items are at least `var.$touch-target-size` in height on mobile?

robertknight commented on src/sidebar/components/datalist.js:
```diff
@@ -0,0 +1,100 @@
+  id,
+  itemPrefixId,
+  list,
+  listFormatter = item => item,
```
Note that since default arguments are evaluated on every function call in JavaScript, `listFormatter` will be different every time if not passed as a prop, this will invalidate the `useMemo` dependencies below. A way to avoid that is to hoist the value to a top-level constant:

```js
const defaultListFormatter = item => item;

function Datalist({
...,
listFormatter = defaultListFormatter,
...
})
```

LMS007 commented on src/sidebar/components/test/datalist-test.js:
```diff
@@ -0,0 +1,165 @@
+import { createElement } from 'preact';
+
+import Datalist from '../datalist';
+import { $imports } from '../tag-editor';
```
good catch!

LMS007 commented:
After doing some exploring about that cmd+option space issue, I decided to pivot the combobox model to one that fits our UI a bit better.

https://hypothes-is.slack.com/archives/C1M8NH76X/p1579814401056000

LMS007 commented on src/sidebar/components/test/tag-editor-test.js:
```diff
@@ -1,18 +1,22 @@
 
   function createComponent(props) {
+    container = document.createElement('div');
+    document.body.appendChild(container);
```
I think I overlooked this, I will use an array

robertknight commented:
This is superseded by https://github.com/hypothesis/client/pull/1712 and https://github.com/hypothesis/client/pull/1714. On that basis I guess we can close this now @LMS007?
